apps/language_server/test/experimental/source_file_test.exs

```diff
@@ -0,0 +1,771 @@
+    end
+
+    test "basic delete" do
+      orig = "foooo\nbar\nbaz"
+
+      assert {:ok, source} =
+               run_changes(orig, [
+                 %{
+                   "text" => "",
+                   "range" => range_create(1, 0, 1, 3)
+                 }
+               ])
+
+      assert "foooo\n\nbaz" = text(source)
+    end
+
+    test "multi-line delete" do
+      orig = "foooo\nbar\nbaz"
+
+      assert {:ok, source} =
+               run_changes(orig, [
+                 %{
+                   "text" => "",
+                   "range" => range_create(0, 5, 1, 3)
+                 }
+               ])
+
+      assert "foooo\nbaz" == text(source)
+    end
+
+    test "single character replace" do
+      orig = "foooo\nbar\nbaz"
+
+      assert {:ok, source} =
+               run_changes(orig, [
+                 %{
+                   "text" => "z",
+                   "range" => range_create(1, 2, 1, 3)
+                 }
+               ])
+
+      assert "foooo\nbaz\nbaz" == text(source)
+    end
+
+    test "multi-character replace" do
+      orig = "foo\nbar"
+
+      assert {:ok, source} =
+               run_changes(orig, [
+                 %{
+                   "text" => "foobar",
+                   "range" => range_create(1, 0, 1, 3)
+                 }
+               ])
+
+      assert "foo\nfoobar" == text(source)
+    end
+
+    test "windows line endings are preserved in document" do
+      orig = "foooo\r\nbar\rbaz"
+
+      assert {:ok, source} =
+               run_changes(orig, [
+                 %{
+                   "text" => "z",
+                   "range" => range_create(1, 2, 1, 3)
+                 }
+               ])
+
+      assert "foooo\r\nbaz\rbaz" == text(source)
+    end
+
+    test "windows line endings are preserved in inserted text" do
+      orig = "foooo\nbar\nbaz"
+
+      assert {:ok, source} =
+               run_changes(orig, [
+                 %{
+                   "text" => "z\r\nz\rz",
+                   "range" => range_create(1, 2, 1, 3)
+                 }
+               ])
+
+      assert "foooo\nbaz\r\nz\rz\nbaz" == text(source)
+    end
+
+    test "utf8 glyphs are preserved in document" do
+      orig = "foooo\nb🏳️‍🌈r\nbaz"
+
+      assert {:ok, source} =
+               run_changes(orig, [
+                 %{
+                   "text" => "z",
+                   "range" => range_create(1, 7, 1, 8)
+                 }
+               ])
+
+      assert "foooo\nb🏳️‍🌈z\nbaz" == text(source)
+    end
+
+    test "utf8 glyphs are preserved in inserted text" do
+      orig = "foooo\nbar\nbaz"
+
+      assert {:ok, source} =
+               run_changes(orig, [
+                 %{
+                   "text" => "z🏳️‍🌈z",
+                   "range" => range_create(1, 2, 1, 3)
+                 }
+               ])
+
+      assert "foooo\nbaz🏳️‍🌈z\nbaz" == text(source)
+    end
+
+    test "invalid update range - before the document starts -> before the document starts" do
```

Comment:
I think we should keep the behavior aligned with https://github.com/microsoft/vscode-languageserver-node/blob/df9ea01af578e02d910bf0fb4c7d4edcd94d4e55/textDocument/src/test/textdocument.test.ts#L379

Reply:
I'll do it, but those seem like extremely poor tests to me, they expressly violate the spec for what the position can contain
(https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#position says both line and character are uinteger which is defined here https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#uinteger as having a range from 0 to 2^31 - 1)
and they introduce undefined behavior. Nowhere in the spec does it say that positions can be negative (in fact, it says the opposite) and it also doesn't say what to do with them if they are. Seeing negative positions indicates a wildly out of spec client.

I'll push a commit, but I beg you to reconsider. Following a spec is better than following microsoft's slapdash tests.

Reply:
Oh, nice catch @scohen , I missed that the type is `uinteger`. Then let's drop that tests. The only specified behavior is for `character` greater than line length.